accepted/2022/FunctionPointers.md

```diff
@@ -0,0 +1,431 @@
+# Function Pointers
+#### Steve Harter
+#### December 13, 2022
+
+# Background
+For additional context, see:
+- The [original feature ask](https://github.com/dotnet/runtime/issues/11354)
+- The [MetadataLoadContext feature ask](https://github.com/dotnet/runtime/issues/43791)
+- The [7.0 introspection attempt](https://github.com/dotnet/runtime/pull/71516)
+- [C# function pointer spec](https://learn.microsoft.com/dotnet/csharp/language-reference/proposals/csharp-9.0/function-pointers)
+
+Function pointers, which have been in ECMA-335 and used by the CLR since the beginning were only recently supported by C# in C# v9.0 ad .NET v6. However, the .NET type system has not been updated to directly support them and they are currently exposed as the `IntPtr` type:
+```cs
+// using typeof:
+Type type = typeof(delegate*<int, bool>); // System.IntPtr (currently)
+
+// using Type.GetType():
+delegate*<int, bool> fn = &MyFunctionPointer;
+type = fn.GetType(); // System.IntPtr (currently)
+
+static bool MyFunctionPointer(int n) { return n == 42; }
+```
+
+Reflection also returns `IntPtr` for both runtime-reflection through `FieldInfo\ParameterInfo\PropertyInfo` and static-reflection through `MetadataLoadContext`:
+```cs
+  FieldInfo fi = typeof(MyClass).GetField("_fn");
+  Type type = fi.FieldType; // System.IntPtr (currently)
+
+  unsafe class MyClass
+  {
+      public delegate*<int, bool> _fn; // A field as a function pointer
+  }
+```
+
+Treating function pointers as `IntPtr` essentially makes reflection introspection impossible, blocking scenarios that care about function pointers. The proposal is to change from `IntPtr` to `Type` which is a **breaking change**:
+```cs
+Type type = typeof(delegate*<int, bool>); // System.Type (proposed design)
+bool isFunctionPointer = type.IsFunctionPointer; // A new method which returns 'true' here
+```
```

Comment:
Would it be possible to touch on the implications on castability? We're not just changing the reflection, but also how things are represented within the runtime.

This:

```csharp
unsafe
{
    object a = new delegate* unmanaged[MemberFunction]<void>[1];
    Console.WriteLine(a is delegate* unmanaged<void>[]);
    Console.WriteLine(a is delegate* unmanaged[SuppressGCTransition]<void>[]);
    Console.WriteLine(a is delegate* unmanaged[MemberFunction, SuppressGCTransition]<void>[]);

    var b = new delegate* unmanaged[MemberFunction]<void>[1];
    Console.WriteLine(b is delegate* unmanaged<void>[]);
    Console.WriteLine(b is delegate* unmanaged[SuppressGCTransition]<void>[]);
    Console.WriteLine(b is delegate* unmanaged[MemberFunction, SuppressGCTransition]<void>[]);
}
```

Currently prints:

```
True
True
True
False
False
False
```

I assume with this work we're going to make it print 6x False. (Notice how C# is trying to be clever but hitting a C#/runtime type system difference.)

For this one:

```csharp
unsafe
{
    object a = new delegate*<ref int, void>[1];
    Console.WriteLine(a is delegate*<out int, void>[]);
    Console.WriteLine(a is delegate*<in int, void>[]);

    var b = new delegate*<ref int, void>[1];
    Console.WriteLine(b is delegate*<out int, void>[]);
    Console.WriteLine(b is delegate*<in int, void>[]);
}
```

We'll keep the same output:

```
True
True
False
False
```

I.e. runtime type system will disagree with the C# type system, the same way how they disagree about e.g. array variance:

```csharp
Console.WriteLine(new int[1] is uint[]);
Console.WriteLine(((object)new int[1]) is uint[]);
```

Reply:
Thanks for the examples especially the array variance one.

I added a section called "Type Identity" - take a look at that especially now since the design doesn't include calling conventions for non-signature types but does include the invisible "CallKind".